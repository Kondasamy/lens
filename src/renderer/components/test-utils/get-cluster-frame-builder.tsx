/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { DiContainer } from "@ogre-tools/injectable";
import type { LensRendererExtension } from "../../../extensions/lens-renderer-extension";
import rendererExtensionsInjectable from "../../../extensions/renderer-extensions.injectable";
import currentlyInClusterFrameInjectable from "../../routes/currently-in-cluster-frame.injectable";
import { extensionRegistratorInjectionToken } from "../../../extensions/extension-loader/extension-registrator-injection-token";
import { computed } from "mobx";
import { renderFor } from "./renderFor";
import observableHistoryInjectable from "../../navigation/observable-history.injectable";
import currentRouteComponentInjectable from "../../routes/current-route-component.injectable";
import { Sidebar } from "../layout/sidebar";
import React from "react";
import { Router } from "react-router";
import { Observer } from "mobx-react";
import subscribeStoresInjectable from "../../kube-watch-api/subscribe-stores.injectable";
import allowedResourcesInjectable from "../../../common/cluster-store/allowed-resources.injectable";
import type { RenderResult } from "@testing-library/react";

interface Options {
  di: DiContainer;
  extensions: LensRendererExtension[];
}

type Callback = () => void;

export interface ClusterFrameBuilder {
  beforeSetups: (callback: Callback) => ClusterFrameBuilder;
  beforeRender: (callback: Callback) => ClusterFrameBuilder;
  render: () => Promise<RenderResult>;
}

export const getClusterFrameBuilder = ({ di, extensions = [] }: Options) => {
  const beforeSetupsCallbacks: Callback[] = [];
  const beforeRenderCallbacks: Callback[] = [];

  di.override(subscribeStoresInjectable, () => () => () => {});

  di.override(rendererExtensionsInjectable, () => computed(() => extensions));

  di.override(currentlyInClusterFrameInjectable, () => true);

  di.override(allowedResourcesInjectable, () =>
    computed(() => new Set<string>()),
  );

  const builder: ClusterFrameBuilder = {
    beforeSetups(callback: () => void) {
      beforeSetupsCallbacks.push(callback);

      return builder;
    },

    beforeRender(callback: () => void) {
      beforeRenderCallbacks.push(callback);

      return builder;
    },

    async render() {
      beforeSetupsCallbacks.forEach(callback => callback());

      await di.runSetups();

      const extensionRegistrators = di.injectMany(
        extensionRegistratorInjectionToken,
      );

      await Promise.all(
        extensions.flatMap((extension) =>
          extensionRegistrators.map((registrator) => registrator(extension, 1)),
        ),
      );

      const render = renderFor(di);

      const history = di.inject(observableHistoryInjectable);
      const currentRouteComponent = di.inject(currentRouteComponentInjectable);

      beforeRenderCallbacks.forEach(callback => callback());

      return render(
        <Router history={history}>
          <Sidebar />

          <Observer>
            {() => {
              const Component = currentRouteComponent.get();

              if (!Component) {
                return null;
              }

              return <Component />;
            }}
          </Observer>
        </Router>,
      );
    },
  };

  return builder;
};
