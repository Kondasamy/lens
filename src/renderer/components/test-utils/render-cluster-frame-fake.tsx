/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { DiContainer } from "@ogre-tools/injectable";
import type { LensRendererExtension } from "../../../extensions/lens-renderer-extension";
import rendererExtensionsInjectable from "../../../extensions/renderer-extensions.injectable";
import currentlyInClusterFrameInjectable from "../../routes/currently-in-cluster-frame.injectable";
import type { Cluster } from "../../../common/cluster/cluster";
import hostedClusterInjectable from "../../../common/cluster-store/hosted-cluster.injectable";
import { extensionRegistratorInjectionToken } from "../../../extensions/extension-loader/extension-registrator-injection-token";
import { computed } from "mobx";
import { renderFor } from "./renderFor";
import observableHistoryInjectable from "../../navigation/observable-history.injectable";
import currentRouteComponentInjectable from "../../routes/current-route-component.injectable";
import { Sidebar } from "../layout/sidebar";
import React from "react";
import { Router } from "react-router";
import { Observer } from "mobx-react";
import subscribeStoresInjectable from "../../kube-watch-api/subscribe-stores.injectable";

interface Options {
  di: DiContainer;
  extensions: LensRendererExtension[];
}

export const renderClusterFrameFake = async ({ di, extensions = [] }: Options) => {
  di.override(subscribeStoresInjectable, () => () => () => {});

  di.override(rendererExtensionsInjectable, () => computed(() => extensions));

  di.override(currentlyInClusterFrameInjectable, () => true);

  const clusterStub = { allowedResources: [] } as Cluster;

  di.override(hostedClusterInjectable, () => clusterStub);

  await di.runSetups();

  const extensionRegistrators = di.injectMany(
    extensionRegistratorInjectionToken,
  );

  await Promise.all(
    extensions.flatMap((extension) =>
      extensionRegistrators.map((registrator) => registrator(extension, 1)),
    ),
  );

  const render = renderFor(di);

  const history = di.inject(observableHistoryInjectable);
  const currentRouteComponent = di.inject(currentRouteComponentInjectable);

  return () => render(
    <Router history={history}>
      <Sidebar />

      <Observer>
        {() => {
          const Component = currentRouteComponent.get();

          if (!Component) {
            return null;
          }

          return <Component />;
        }}
      </Observer>
    </Router>,
  );
};
