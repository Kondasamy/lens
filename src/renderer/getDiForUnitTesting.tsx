/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import glob from "glob";
import {
  join,
  memoize,
  split,
  noop,
  slice,
  filter,
  last,
  map,
  tap,
} from "lodash/fp";
import { createContainer } from "@ogre-tools/injectable";
import { setLegacyGlobalDiForExtensionApi } from "../extensions/as-legacy-globals-for-extension-api/legacy-global-di-for-extension-api";
import getValueFromRegisteredChannelInjectable from "./app-paths/get-value-from-registered-channel/get-value-from-registered-channel.injectable";
import loggerInjectable from "../common/logger.injectable";
import fsInjectable from "../common/fs/fs.injectable";
import type fse from "fs-extra";
import { pipeline } from "@ogre-tools/fp";

export const getDiForUnitTesting = (
  { doGeneralOverrides } = { doGeneralOverrides: false },
) => {
  const di = createContainer();

  setLegacyGlobalDiForExtensionApi(di);

  for (const filePath of getInjectableFilePaths()) {
    const injectableInstance = require(filePath).default;

    di.register({
      ...injectableInstance,
      aliases: [injectableInstance, ...(injectableInstance.aliases || [])],
    });
  }

  di.preventSideEffects();

  if (doGeneralOverrides) {
    di.override(getValueFromRegisteredChannelInjectable, () => () => undefined);

    di.override(fsInjectable, () => getFsFake());

    di.override(loggerInjectable, () => ({
      warn: noop,
      debug: noop,
      log: noop,
      error: (...args: any) => console.error(...args),
      info: noop,
    }));
  }

  return di;
};

const getInjectableFilePaths = memoize(() => [
  ...glob.sync("./**/*.injectable.{ts,tsx}", { cwd: __dirname }),
  ...glob.sync("../common/**/*.injectable.{ts,tsx}", { cwd: __dirname }),
  ...glob.sync("../extensions/**/*.injectable.{ts,tsx}", { cwd: __dirname }),
]);

const getFsFake = () => {
  const state = new Map();

  const readFile = (filePath: string) => {
    const fileContent = state.get(filePath);

    if (!fileContent) {
      const existingFilePaths = [...state.keys()].join('", "');

      throw new Error(
        `Tried to access file ${filePath} which does not exist. Existing file paths are: "${existingFilePaths}"`,
      );
    }

    return Promise.resolve(Buffer.from(fileContent));
  };

  return {
    ensureDir: () => Promise.resolve(),

    writeJson: (filePath: string, contents: object) => {
      state.set(filePath, JSON.stringify(contents));

      return Promise.resolve();
    },

    readJson: async (filePath: string) => {
      const fileContent = await readFile(filePath);

      return JSON.parse(fileContent.toString());
    },

    readFile,

    readdir: (directoryPath: string) => {
      const fileNames = pipeline(
        [...state.keys()],

        filter((x) =>
          pipeline(
            x,
            split("/"),
            slice(0, -1),
            join("/"),
          ),
        ),

        tap(console.log),

        map((x) => pipeline(x, split("/"), last)),
      );

      return Promise.resolve(fileNames);
    },
  } as unknown as typeof fse;
};
