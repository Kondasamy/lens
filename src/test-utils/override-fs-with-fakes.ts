/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { DiContainer } from "@ogre-tools/injectable";
import readFileInjectable from "../common/fs/read-file.injectable";
import writeJsonFileInjectable from "../common/fs/write-json-file.injectable";
import readJsonFileInjectable from "../common/fs/read-json-file.injectable";
import pathExistsInjectable from "../common/fs/path-exists.injectable";

export const overrideFsWithFakes = (di: DiContainer) => {
  const state = new Map();

  const readFile = readFileFor(state);

  const overrides = [
    {
      injectableToOverride: readFileInjectable,

      instantiateFake: () => readFile,
    },

    {
      injectableToOverride: writeJsonFileInjectable,

      instantiateFake: () => (filePath: string, contents: object) => {
        state.set(filePath, JSON.stringify(contents));

        return Promise.resolve();
      },
    },

    {
      injectableToOverride: readJsonFileInjectable,

      instantiateFake: () => async (filePath: string) => {
        const fileContent = await readFile(filePath);

        return JSON.parse(fileContent.toString());
      },
    },

    {
      injectableToOverride: pathExistsInjectable,

      instantiateFake: () => async (filePath: string) =>
        Promise.resolve(state.has(filePath)),
    },
  ];

  overrides.forEach(({ injectableToOverride, instantiateFake }) => {
    di.override(injectableToOverride, instantiateFake);
  });
};

const readFileFor = (state: Map<string, Buffer>) => (filePath: string) => {
  const fileContent = state.get(filePath);

  if (!fileContent) {
    const existingFilePaths = [...state.keys()].join('", "');

    throw new Error(
      `Tried to access file ${filePath} which does not exist. Existing file paths are: "${existingFilePaths}"`,
    );
  }

  return Promise.resolve(Buffer.from(fileContent));
};
