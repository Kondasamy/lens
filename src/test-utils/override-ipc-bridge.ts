/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { DiContainer } from "@ogre-tools/injectable";
import getValueFromChannelInjectable from "../renderer/channel/get-value-from-channel.injectable";
import registerChannelInjectable from "../main/app-paths/register-channel/register-channel.injectable";
import asyncFn from "@async-fn/jest";
import type { SendToViewArgs } from "../main/start-main-application/lens-window/application-window/lens-window-injection-token";
import sendToChannelInElectronBrowserWindowInjectable from "../main/start-main-application/lens-window/application-window/send-to-channel-in-electron-browser-window.injectable";
import { isEmpty } from "lodash/fp";
import enlistMessageChannelListenerInjectableInRenderer from "../renderer/channel/channel-listeners/enlist-message-channel-listener.injectable";
import enlistMessageChannelListenerInjectableInMain from "../main/channel/channel-listeners/enlist-message-channel-listener.injectable";
import enlistRequestChannelListenerInjectableInMain from "../main/channel/channel-listeners/enlist-request-channel-listener.injectable";
import sendToMainInjectable from "../renderer/channel/send-to-main.injectable";
import type { RequestChannel } from "../common/channel/request-channel-injection-token";

export const overrideIpcBridge = ({
  rendererDi,
  mainDi,
}: {
  rendererDi: DiContainer;
  mainDi: DiContainer;
}) => {
  const fakeRequestChannelMap = new Map<
    string,
    { promise: Promise<any>; resolve: (arg0: any) => Promise<void> }
  >();

  const requestChannelListenerFakesForMain = {
    set: <TChannel extends RequestChannel<unknown, unknown>>(
      channel: TChannel,
      callback: () => TChannel["_responseSignature"],
    ) => {
      const id = channel.id;

      if (!fakeRequestChannelMap.has(id)) {
        const mockInstance = asyncFn();

        fakeRequestChannelMap.set(id, {
          promise: mockInstance(),
          resolve: mockInstance.resolve,
        });
      }

      return fakeRequestChannelMap.get(id)?.resolve(callback);
    },

    get: <TChannel extends RequestChannel<unknown, unknown>>(channel: TChannel) => {
      const id = channel.id;

      if (!fakeRequestChannelMap.has(id)) {
        const mockInstance = asyncFn();

        fakeRequestChannelMap.set(id, {
          promise: mockInstance(),
          resolve: mockInstance.resolve,
        });
      }

      return fakeRequestChannelMap.get(id)?.promise;
    },
  };

  rendererDi.override(
    getValueFromChannelInjectable,

    () => async (channel) => {
      const callback = await requestChannelListenerFakesForMain.get(channel);

      return callback();
    },
  );

  mainDi.override(registerChannelInjectable, () => (channel, listener) => {
    requestChannelListenerFakesForMain.set(channel, listener);
  });

  const messageChannelListenerFakesForRenderer = new Map<
    string,
    ((...args: any[]) => void)[]
  >();

  mainDi.override(
    sendToChannelInElectronBrowserWindowInjectable,

    () =>
      (browserWindow, { channel: channelName, data = [] }: SendToViewArgs) => {
        const listeners = messageChannelListenerFakesForRenderer.get(channelName) || [];

        if (isEmpty(listeners)) {
          throw new Error(
            `Tried to send message to channel "${channelName}" but there where no listeners. Current channels with listeners: "${[
              ...messageChannelListenerFakesForRenderer.keys(),
            ].join('", "')}"`,
          );
        }

        listeners.forEach((handle) => handle(...data));
      },
  );

  const messageChannelListenerFakesForMain = new Map<
    string,
    ((...args: any[]) => void)[]
  >();

  const requestChannelListenerFakesForMain = new Map<
    string,
    ((...args: any[]) => any)[]
  >();

  rendererDi.override(
    enlistMessageChannelListenerInjectableInRenderer,

    () =>
      ({ channel, handler: newListener }) => {
        const oldListeners = messageChannelListenerFakesForRenderer.get(channel.id) || [];

        messageChannelListenerFakesForRenderer.set(channel.id, [...oldListeners, newListener]);

        return () => {};
      },
  );

  rendererDi.override(sendToMainInjectable, () => (channelId, message) => {
    const listeners = messageChannelListenerFakesForMain.get(channelId) || [];

    if (isEmpty(listeners)) {
      throw new Error(
        `Tried to send message to channel "${channelId}" but there where no listeners. Current channels with listeners: "${[
          ...messageChannelListenerFakesForMain.keys(),
        ].join('", "')}"`,
      );
    }

    listeners.forEach((handle) => handle(message));
  });

  mainDi.override(
    enlistMessageChannelListenerInjectableInMain,

    () =>
      ({ channel, handler: newListener }) => {
        const oldListeners = messageChannelListenerFakesForMain.get(channel.id) || [];

        messageChannelListenerFakesForMain.set(channel.id, [...oldListeners, newListener]);

        return () => {};
      },
  );

  mainDi.override(
    enlistRequestChannelListenerInjectableInMain,

    () =>
      ({ channel, handler }) => {
        // const existingHandles = mainIpcFakeHandles.get(channel.id) || [];

        // mainIpcFakeHandles.set(channel.id, [...existingHandles, handler]);

        return () => {};
      },
  );
};
