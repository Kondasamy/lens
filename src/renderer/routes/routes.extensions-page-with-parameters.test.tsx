/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { fireEvent, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import type { DiContainer } from "@ogre-tools/injectable";
import { getDiForUnitTesting } from "../getDiForUnitTesting";
import React from "react";
import { computed, runInAction } from "mobx";
import rendererExtensionsInjectable from "../../extensions/renderer-extensions.injectable";
import { LensRendererExtension } from "../../extensions/lens-renderer-extension";
import observableHistoryInjectable from "../navigation/observable-history.injectable";
import { createMemoryHistory, MemoryHistory } from "history";
import { createObservableHistory } from "mobx-observable-history";
import currentlyInClusterFrameInjectable from "./currently-in-cluster-frame.injectable";
import directoryForUserDataInjectable from "../../common/app-paths/directory-for-user-data/directory-for-user-data.injectable";

import type { PageRegistration } from "../../extensions/registries";
import { renderFor } from "../components/test-utils/renderFor";
import { Observer } from "mobx-react";
import currentRouteInjectable from "./current-route.injectable";
import type { RenderResult } from "@testing-library/react";

describe("routes - extensions - page with parameters", () => {
  let historyFake: MemoryHistory;
  let di: DiContainer;
  let rendered: RenderResult;

  beforeEach(async () => {
    di = getDiForUnitTesting({ doGeneralOverrides: true });

    di.override(
      directoryForUserDataInjectable,
      () => "some-directory-for-user-data",
    );

    const extension = new TestExtension({
      id: "some-extension-id",
      clusterPages: [],

      globalPages: [
        {
          components: {
            Page: ({ params }) => {
              return (
                <div>
                  <span data-testid="test-value">
                    {params.someNonNormalizedParameter.get()}
                  </span>

                  {params.someNormalizedParameter.get()}

                  <button
                    data-testid="some-button"
                    onClick={() =>
                      params.someNonNormalizedParameter.set(
                        "some-changed-value",
                      )
                    }
                  />
                </div>
              );
            },
          },

          params: {
            someNonNormalizedParameter: "some-initial-value",

            someNormalizedParameter: {
              defaultValue: "some-initial-value",
              stringify: (asd) =>
                asd === "uusi-arvo" && "uusi-stringified-arvo",
              parse: (asd) =>
                asd === "uusi-stringified-arvo" && "some-changed-value",
            },
          },
        },
      ],
    });

    di.override(rendererExtensionsInjectable, () =>
      computed(() => [extension]),
    );

    historyFake = createMemoryHistory();
    const observableHistory = createObservableHistory(historyFake);

    di.override(observableHistoryInjectable, () => observableHistory);
    di.override(currentlyInClusterFrameInjectable, () => false);

    await di.runSetups();

    const render = renderFor(di);

    const currentRoute = di.inject(currentRouteInjectable);

    rendered = render(
      <Observer>
        {() => {
          const route = currentRoute.get();

          if (!route) {
            return <div data-testid="no-active-route" />;
          }

          return <route.Component />;
        }}
      </Observer>,
    );
  });

  describe("when accessing route", () => {
    beforeEach(() => {
      runInAction(() => {
        historyFake.replace("/extension/some-extension-id");
      });
    });

    it("has value", () => {
      expect(screen.getByTestId("test-value")).toHaveTextContent(
        "some-initial-value",
      );
    });

    it("renders", () => {
      expect(rendered.container).toMatchSnapshot();
    });

    describe("when page parameter is changed", () => {
      beforeEach(() => {
        const button = rendered.getByTestId("some-button");

        fireEvent.click(button);
      });

      it("updates URL", () => {
        const url = historyFake.location.pathname + historyFake.location.search;

        expect(url).toBe(
          "/extension/some-extension-id?some-extension-id%3AsomeNonNormalizedParameter=some-changed-value",
        );
      });

      it("has value", () => {
        expect(rendered.getByTestId("test-value")).toHaveTextContent(
          "some-changed-value",
        );
      });

      it("renders", () => {
        expect(rendered.container).toMatchSnapshot();
      });
    });
  });
});

class TestExtension extends LensRendererExtension {
  constructor({
    id,
    globalPages,
    clusterPages,
  }: {
    id: string;
    globalPages: PageRegistration[];
    clusterPages: PageRegistration[];
  }) {
    super({
      id,
      absolutePath: "irrelevant",
      isBundled: false,
      isCompatible: false,
      isEnabled: false,
      manifest: { name: id, version: "some-version" },
      manifestPath: "irrelevant",
    });

    this.globalPages = globalPages;
    this.clusterPages = clusterPages;
  }
}
