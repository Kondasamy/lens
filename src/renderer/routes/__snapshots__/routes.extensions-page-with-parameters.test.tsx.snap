// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes - extensions - page with parameters when accessing route renders 1`] = `
<div>
  <div>
    <span
      data-testid="test-value"
    >
      some-initial-value
    </span>
    <button
      data-testid="some-button"
    />
  </div>
</div>
`;

exports[`routes - extensions - page with parameters when accessing route when page parameter is changed renders 1`] = `
<div>
  <div>
    <span
      data-testid="test-value"
    >
      some-changed-value
    </span>
    <button
      data-testid="some-button"
    />
  </div>
</div>
`;
